src/analysis/powerset-lattice-impl.h

```diff
@@ -5,61 +5,115 @@
+  return bitvector.at(index);
 }
 
-template<size_t N>
 inline LatticeComparison
-BitsetPowersetLattice<N>::compare(const BitsetPowersetLattice<N>& left,
-                                  const BitsetPowersetLattice<N>& right) {
-  size_t leftCount = left.value.count();
-  size_t rightCount = right.value.count();
-
-  // If left has more elements, left might be a superset of right.
-  if (leftCount > rightCount) {
-    if ((left.value | right.value) == left.value) {
-      return GREATER;
+FinitePowersetLattice::compare(const FinitePowersetLattice& left,
+                               const FinitePowersetLattice& right) {
+  // Both must be from the powerset lattice of the same set.
+  assert(left.bitvector.size() == right.bitvector.size());
+
+  // True in left, false in right.
+  bool leftNotRight = false;
+
+  // True in right, false in left.
+  bool rightNotLeft = false;
+
+  if (left.trues > right.trues) {
+    // If there are more elements in the left, some must not be in the right.
+    leftNotRight = true;
+  } else if (right.trues > left.trues) {
+    // If there are more elements in the right, some must not be in the left.
+    rightNotLeft = true;
+  }
+
+  size_t size = left.bitvector.size();
+
+  for (size_t i = 0; i < size; ++i) {
+    if (left.bitvector[i] && !right.bitvector[i]) {
+      leftNotRight = true;
+    } else if (right.bitvector[i] && !left.bitvector[i]) {
+      rightNotLeft = true;
     }
-    // If right has more elements, right might be a superset of left.
-  } else if (leftCount < rightCount) {
-    if ((left.value | right.value) == right.value) {
-      return LESS;
+
+    // We can end early if we know neither is a subset of the other.
+    if (leftNotRight && rightNotLeft) {
+      return NO_RELATION;
+    }
+  }
+
+  if (!leftNotRight) {
+    if (!rightNotLeft) {
+      return EQUAL;
     }
-  } else if (left.value == right.value) {
-    return EQUAL;
+    return LESS;
+  } else if (!rightNotLeft) {
+    return GREATER;
   }
 
   return NO_RELATION;
 }
 
-// Implement LUB as an OR of the two bitsets.
-template<size_t N>
-inline void BitsetPowersetLattice<N>::getLeastUpperBound(
-  const BitsetPowersetLattice<N>& right) {
-  value |= right.value;
+// Least upper bound is implemented as a logical OR between the bitvectors on
+// both sides.
+inline bool
+FinitePowersetLattice::getLeastUpperBound(const FinitePowersetLattice& right) {
+  // Both must be from powerset lattice of the same set.
+  assert(right.bitvector.size() == bitvector.size());
+
+  bool modified = false;
+  for (size_t i = 0; i < bitvector.size(); ++i) {
+    if (!bitvector.at(i) && right.bitvector.at(i)) {
+      bitvector.at(i) = true;
+      trues++;
+      modified = true;
+    }
+  }
+
+  return modified;
 }
 
-template<size_t N>
-inline void BitsetPowersetLattice<N>::print(std::ostream& os) {
-  os << value;
+inline void FinitePowersetLattice::print(std::ostream& os) {
+  for (auto it = bitvector.rbegin(); it != bitvector.rend(); ++it) {
```

Comment:
Why print this in reverse?

Reply:
The bitset vector printed out the values in reverse, so I did it here for backward compatibility with the tests. But yes, I should probably change the test instead.

Reply:
current solution lgtm.